### PR TITLE
NC added suggested code to detect_timeout.py

### PR DESCRIPTION
**Pull request by clarkenj**
Have not actually edited lines that were there - can decide how we want to proceed

### CODE REVIEW

File: scripts/detect_timeout.py [hunk 1]
```diff
@@ -1,7 +1,45 @@
 import argparse
 import re
+from nltk import word_tokenize
```
Comment (htwangtw) [inline scripts/detect_timeout.py:3]: I am intentionally using only packages that comes with base installation. I would avoid heaving dependency for such small job.

Comment (htwangtw) [reply scripts/detect_timeout.py:3]: Eh looks like this import was not even used

File: scripts/detect_timeout.py [hunk 2]
```diff
@@ -1,7 +1,45 @@
...
+        #loop through tokens and see if they correspond to common error terms. If so, add file name to list
+        for token in split_line:
+            if token == 'out-of-memory':
+                out_memory_list.append(file.stem)
+
+            if token == 'LIMIT': 
+                time_limit_list.append(file.stem)
+
+            if token == 'FunctionalSummaryInputSpec':
+                trait_list.append(file.stem)
```
Comment (htwangtw) [inline scripts/detect_timeout.py:31]: It's great that you have identified a lot of the keywords! 
Let's work on integrating all these into the `check_timeout` function below. We can certainly rename the function to `check_error_logs` or something.

Comment (htwangtw) [reply scripts/detect_timeout.py:31]: You will need to integrate this block to line 62. It can be a bunch of if statements for now.

**Review by htwangtw — COMMENTED** [review]
Thanks for identifying the failed cases! 

I probably should have clarified that when seeing a PR, what people expect is that the changes will run with the existing code. 
I can do a separate PR to integrate your code, but I would prefer you to complete it so you understand the full process.

Try integrate the code into the existing function, so the tool can run as it is. 

Here's explanations of the key files used in this script. I hope this give you a better idea to read the code and how you should complete this PR.

Takes one input, which is the path to a fmriprep slurm directory.
It checks three things:
1. subjects successfully finished fMRIPrep: identified through the presence of the `sub-*` directory under the fmriprep derivative directory. Since in giga processing 2 we are using a standardised name, I can safely assume `"fmriprep*/sub-*"` is a suitable pattern to search. Let's call this group set `completed_subjects`. 
2. subjects submitted: similar logic as above, I used pattern `".slurm/smriprep_sub-*"`. Let's call this group set `submitted_subjects`.
3. subject error logs: I did not index this directly. They are only accessed if the subject is present in the set difference of set `completed_subjects` and set `submitted_subjects`. Let's call this group set `failed_subjects`

We will search the keywords in the error logs of all subjects present in set `failed_subjects`.

**Comment by clarkenj** [comment]
Thanks for the comment, that makes it a lot clearer what you wanted. But I'm still not sure what the outputs of the code would be, other than what you had? I.e. what's the goal of me editing it?

**Comment by htwangtw** [comment]
> what's the goal of me editing it?

We will have one script to detect multiple failed cases. 
If you feel this is too much I can finish it off and see if this is useful for you as a tool to access future datasets

**Comment by clarkenj** [comment]
Ok sounds good. I'm happy to do it. My previous slurm folder is now archived so may take a bit longer

**Comment by clarkenj** [comment]
@htwangtw I have edited the script, sorry for the delay. I made the following changes:

- edited function to accept multiple replacements to strings in the .slurm file
- added code to find subjects who failed with `out-of-memory` (...'due to may have been killed by the cgroup out-of-memory handler')
- for these subjects a) add flag `--mem-mb 11000` (slightly less than --mem-per-cpu) - this is hard coded, but may also change given new approach for resource use, and b) increase wall time
- instead of editing the .slurm files in-place, creates a new folder with the modified scripts. This was because for adni I had a lot of subjects fail, and it would take a while to submit them one by one

Let me know what you think!

File: scripts/detect_timeout.py [hunk 3]
```diff
@@ -24,17 +24,23 @@ def check_timeout(args):
...
+        except FileNotFoundError:
+            pass
+
+        try:
+            with open(fmriprep_slurm_output / f"smriprep_{s}.out") as f:
+                txt=f.read()
+                if re.search(r'\bnipype.workflow ERROR\b', txt):
+                    workflow_error_subjects.add(s)
+        except FileNotFoundError:
+            pass
```
Comment (htwangtw) [inline scripts/detect_timeout.py:43]: same as above

File: scripts/detect_timeout.py [hunk 4]
```diff
@@ -1,3 +1,4 @@
+import warnings
```
Comment (clarkenj) [inline scripts/detect_timeout.py]: this should have been done in the add warnings commit, oops

**Comment by htwangtw** [comment]
I am going to do a PR to this PR and make some option more flexible! Well done!

**Comment by clarkenj** [comment]
@htwangtw I just noticed a mistake, the final print message is indented too many times. I am fixing it.

**Comment by htwangtw** [comment]
I have fixed those two things in the PR so I just resolved them

**Pull request merged**